autoload/phpactor/project/assigner.vim

```diff
@@ -0,0 +1,118 @@
+function! phpactor#project#assigner#assignProjectToBuffer(filename)
+  if !exists('g:phpactorLoaded')
+    return
+  endif
+
+  if exists('b:project')
+    return
+  endif
+  let l:project = g:phpactorBufferMatcher.resolveProjectForFile(a:filename)
+  if v:null is l:project
+    return
+  endif
+  let b:project = l:project
+  if g:phpactorBufferMatcher.repository.hasProject(b:project)
+    return
+  endif
+
+  call g:phpactorBufferMatcher.repository.addProject(b:project)
+endfunction
+
+function! phpactor#project#assigner#create(projectRepository, projectRootMarkers, filesystemRootMarkers, initialCwd) abort
+  let l:filesystemRootMarkers = a:filesystemRootMarkers
+  if index(l:filesystemRootMarkers, '/') < 0
+    call add(l:filesystemRootMarkers, '/')
+  endif
+
+  return {
+        \ 'initialCwd': a:initialCwd,
+        \ 'projectRootMarkers': a:projectRootMarkers,
+        \ 'filesystemRootMarkers': l:filesystemRootMarkers,
+        \ 'repository': a:projectRepository,
+        \ 'resolveProjectForFile': function('s:resolveProjectForFile')
+        \ }
+endfunction
+
+function s:resolveProjectForFile(file) dict abort
+  let l:project = self.repository.findProjectContainingFile(a:file)
+
+  if type(l:project) == type({})
+    return l:project
+  endif
+
+  " early version to refactor or change
+  " heredoc are available since vim 8.1.1354
+  let l:message = printf(join([
+        \ 'Phpactor [RPC]',
+        \ 'There is no project enabled for "%s" yet.',
+        \ 'Select the way of assign the project root',
+        \ 'If cancelled, "%s" will be used.'
+        \], "\n"), a:file, self.initialCwd)
+
+  " see phpactor#_path
+  let l:initialDirectory = fnamemodify(a:file, ':p:h')
+  let l:rootDirByMarker = s:searchDirectoryUpwardForRootMarkers(
+        \ l:initialDirectory,
+        \ self.projectRootMarkers,
+        \ self.filesystemRootMarkers
+        \ )
+
+  let l:choices = [
+        \ {
+        \ 'action': { -> self.initialCwd },
+        \ 'message': l:message
+        \ },
+        \ {
+        \ 'action': function('input', ['Enter file path: ', l:initialDirectory, 'file']),
```

Comment:
So you don't want to ask again if it's not a valid path ?
Typo are common even with completion I think we should help the user as much as possible.